### PR TITLE
Add a method to print the stack trace

### PR DESCRIPTION
**Pull request by aaronfranke**
This adds a new method to the error macros that allows printing the stack trace. Add `print_stack_trace();` to any method to print the stack trace at that point. You can also specify if you want to skip a part of the stack trace at the beginning or end, to avoid any noisy lines you don't care about. It's only available in debug builds.

There is no way to print the stack trace in a cross-platform way, so we need platform-specific code. On Mac and Linux, the code to print the stack trace is simple, since there is a handy library that does all the work for us, but we don't have fine control over the output. On Windows, it's more complicated, the code was designed based on [this SO post](https://stackoverflow.com/questions/6205981/windows-c-stack-trace-from-a-running-app) and [this article](https://www.rioki.org/2017/01/09/windows_stacktrace.html). I made the Windows code's output somewhat mimic the Mac/Linux code for consistency between platforms. Also, Windows stack trace printing only works on x86_32 and x86_64 (and technically Itanium, but Godot doesn't run on that).

This work was sponsored by The Mirror, we plan to use this by adding `print_stack_trace(2);` to the `void _err_print_error` method so that we will get stack traces for every error (and the `2` skips the calls to `_err_print_error` itself). This is not included in this PR since it would be spammy, but it's one line away for anyone who wants it.

### CODE REVIEW

**Comment by Zylann** [comment]
How about obtaining the stack trace in a data structure instead of directly printing it? This way it can be routed elsewhere or formatted.

**Comment by Calinou** [comment]
> How about obtaining the stack trace in a data structure instead of directly printing it? This way it can be routed elsewhere or formatted in other ways.

I think this is worth adding as a separate method, but the method for quick printing should remain available for convenience.

**Comment by akien-mga** [comment]
For the record (Aaron is already aware), this partially overlaps with the existing crash handler implementations in `platform/*/crash_handler_*.cpp` so it should likely be refactored to use a common base for both features. The platform specific code should likely be kept in `platform` so it's more modular (and allows implementing it for third-party platforms too, if relevant).

**Comment by aaronfranke** [comment]
Linux print_stack_trace non-verbose:
```
[ 1] godot.linuxbsd.editor.dev.arm64    - initialize_gltf_module(ModuleInitializationLevel) (register_types.cpp:113)
[ 2] godot.linuxbsd.editor.dev.arm64    - initialize_modules(ModuleInitializationLevel) (register_module_types.gen.cpp:146)
[ 3] godot.linuxbsd.editor.dev.arm64    - Main::setup(char const*, int, char**, bool) (main.cpp:2256)
[ 4] godot.linuxbsd.editor.dev.arm64    - main (godot_linuxbsd.cpp:111)
```

Linux print_stack_trace in verbose mode:
```
[ 1] b05834e8c76c (godot.linuxbsd.editor.dev.arm64+484c76c)     - initialize_gltf_module(ModuleInitializationLevel) (/home/aaronfranke/workspace/godot/godot/modules/gltf/register_types.cpp:113)
[ 2] b05834c80304 (godot.linuxbsd.editor.dev.arm64+4640304)     - initialize_modules(ModuleInitializationLevel) (/home/aaronfranke/workspace/godot/godot/modules/register_module_types.gen.cpp:146)
[ 3] b058342f05cc (godot.linuxbsd.editor.dev.arm64+3cb05cc)     - Main::setup(char const*, int, char**, bool) (/home/aaronfranke/workspace/godot/godot/main/main.cpp:2256)
[ 4] b0583426ff10 (godot.linuxbsd.editor.dev.arm64+3c2ff10)     - main (/home/aaronfranke/workspace/godot/godot/platform/linuxbsd/godot_linuxbsd.cpp:111)
```

Linux crash with this PR:
```
================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.7.dev.custom_build (fee7e394adf12a45a85d02b6f07112b3f47ade56)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
Load address: c5c042c20000

[ 1] ef12426bb8f8 - ?? ??:0
[ 2] c5c04750d154 - Node::set_name(StringName const&) at /home/aaronfranke/workspace/godot/godot/scene/main/node.cpp:1442
[ 3] c5c0440f4db4 - initialize_gltf_module(ModuleInitializationLevel) at /home/aaronfranke/workspace/godot/godot/modules/gltf/register_types.cpp:113 (discriminator 2)
[ 4] c5c043ee3014 - initialize_modules(ModuleInitializationLevel) at /home/aaronfranke/workspace/godot/godot/modules/register_module_types.gen.cpp:147
[ 5] c5c043500a30 - Main::setup(char const*, int, char**, bool) at /home/aaronfranke/workspace/godot/godot/main/main.cpp:2256
[ 6] c5c04348127c - main at /home/aaronfranke/workspace/godot/godot/platform/linuxbsd/godot_linuxbsd.cpp:111
[ 7] ef12424184c4 - ?? ??:0
[ 8] ef1242418598 - ?? ??:0
[ 9] c5c04345e5f0 - _start at ??:?
-- END OF C++ BACKTRACE --
================================================================
```

Linux crash without this PR for comparison:

```
================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.7.dev.custom_build (3f63a400fa8f454b09f12201152025c1805b3df5)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
Load address: ae2aa53b0000

[1] f6fa354b18f8 - ?? ??:0
[2] ae2aa9cad390 - Node::set_name(StringName const&) at /home/aaronfranke/workspace/godot/godot/scene/main/node.cpp:1442
[3] ae2aa688d36c - initialize_gltf_module(ModuleInitializationLevel) at /home/aaronfranke/workspace/godot/godot/modules/gltf/register_types.cpp:113 (discriminator 2)
[4] ae2aa667a5cc - initialize_modules(ModuleInitializationLevel) at /home/aaronfranke/workspace/godot/godot/modules/register_module_types.gen.cpp:147
[5] ae2aa5c918ac - Main::setup(char const*, int, char**, bool) at /home/aaronfranke/workspace/godot/godot/main/main.cpp:2256
[6] ae2aa5c0ffd8 - main at /home/aaronfranke/workspace/godot/godot/platform/linuxbsd/godot_linuxbsd.cpp:111
[7] f6fa352184c4 - ?? ??:0
[8] f6fa35218598 - ?? ??:0
[9] ae2aa5bef5b0 - _start at ??:?
-- END OF C++ BACKTRACE --
================================================================
```

Before PR #117060:

<details>

Linux crash with this PR:

```
================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.7.dev.custom_build (bd61868e7c6e81fd29f93caf46295fe661614092)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
Load address: c0a6225e0000

[ 1] e0648b4498f8 - linux-vdso.so.1(__kernel_rt_sigreturn+0) [0xe0648b4498f8] (??:0)
[ 2] c0a62a0b19f4 - Node::set_name(StringName const&) (/home/aaronfranke/workspace/godot/godot/scene/main/node.cpp:1442)
[ 3] c0a626e2f7d0 - initialize_gltf_module(ModuleInitializationLevel) (/home/aaronfranke/workspace/godot/godot/modules/gltf/register_types.cpp:114 (discriminator 2))
[ 4] c0a626c21330 - initialize_modules(ModuleInitializationLevel) (/home/aaronfranke/workspace/godot/godot/modules/register_module_types.gen.cpp:146)
[ 5] c0a62628e5f4 - Main::setup(char const*, int, char**, bool) (/home/aaronfranke/workspace/godot/godot/main/main.cpp:2256)
[ 6] c0a62620ff40 - ./bin/godot.linuxbsd.editor.dev.arm64(main+0x114) [0xc0a62620ff40] (/home/aaronfranke/workspace/godot/godot/platform/linuxbsd/godot_linuxbsd.cpp:111)
[ 7] e0648b1a84c4 - /lib/aarch64-linux-gnu/libc.so.6(+0x284c4) [0xe0648b1a84c4] (??:0)
[ 8] e0648b1a8598 - /lib/aarch64-linux-gnu/libc.so.6(__libc_start_main+0x98) [0xe0648b1a8598] (??:0)
[ 9] c0a6261ed2f0 - ./bin/godot.linuxbsd.editor.dev.arm64(_start+0x30) [0xc0a6261ed2f0] (??:?)
-- END OF C++ BACKTRACE --
================================================================
```

Linux crash without this PR for comparison:

```
================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.7.dev.custom_build (ee713ccb7cbaddf59f8258ed6663461779d00856)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
Load address: c02cfd4c0000

[1] f707f12c18f8 - linux-vdso.so.1(__kernel_rt_sigreturn+0) [0xf707f12c18f8] (??:0)
[2] c02d04fa7658 - Node::set_name(StringName const&) (/home/aaronfranke/workspace/godot/godot/scene/main/node.cpp:1442)
[3] c02d01d077a0 - initialize_gltf_module(ModuleInitializationLevel) (/home/aaronfranke/workspace/godot/godot/modules/gltf/register_types.cpp:113 (discriminator 2))
[4] c02d01afc314 - initialize_modules(ModuleInitializationLevel) (/home/aaronfranke/workspace/godot/godot/modules/register_module_types.gen.cpp:146)
[5] c02d0116fecc - Main::setup(char const*, int, char**, bool) (/home/aaronfranke/workspace/godot/godot/main/main.cpp:2256)
[6] c02d010ee6f4 - ./bin/godot.linuxbsd.editor.dev.arm64(main+0x114) [0xc02d010ee6f4] (/home/aaronfranke/workspace/godot/godot/platform/linuxbsd/godot_linuxbsd.cpp:111)
[7] f707f10284c4 - /lib/aarch64-linux-gnu/libc.so.6(+0x284c4) [0xf707f10284c4] (??:0)
[8] f707f1028598 - /lib/aarch64-linux-gnu/libc.so.6(__libc_start_main+0x98) [0xf707f1028598] (??:0)
[9] c02d010cbf70 - ./bin/godot.linuxbsd.editor.dev.arm64(_start+0x30) [0xc02d010cbf70] (??:?)
-- END OF C++ BACKTRACE --
================================================================
```
</details>

macOS print_stack_trace non-verbose:
```
[ 1] main       - initialize_gltf_module(ModuleInitializationLevel) (in godot.macos.editor.dev.arm64) + 52 (unknown file)
[ 2] main       - initialize_modules(ModuleInitializationLevel) (in godot.macos.editor.dev.arm64) + 144 (unknown file)
[ 3] main       - Main::setup(char const*, int, char**, bool) (in godot.macos.editor.dev.arm64) + 18836 (unknown file)
[ 4] main       - OS_MacOS_NSApp::start_main() (in godot.macos.editor.dev.arm64) + 60 (unknown file)
```

macOS print_stack_trace in verbose mode:
```
[ 1] 1057c8800 (main+1570800)   - initialize_gltf_module(ModuleInitializationLevel) (in godot.macos.editor.dev.arm64) + 52 (unknown file)
[ 2] 1055c1560 (main+1369560)   - initialize_modules(ModuleInitializationLevel) (in godot.macos.editor.dev.arm64) + 144 (unknown file)
[ 3] 104a0ccf8 (main+7b4cf8)    - Main::setup(char const*, int, char**, bool) (in godot.macos.editor.dev.arm64) + 18836 (unknown file)
[ 4] 104957788 (main+6ff788)    - OS_MacOS_NSApp::start_main() (in godot.macos.editor.dev.arm64) + 60 (unknown file)
```

macOS crash with this PR:
```
================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.7.dev.custom_build (cc0a112da8476b7b0651168a9d14cfb3d17f7f08)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
Load address: 100230000

[ 1] 18ec897a4 - 1   libsystem_platform.dylib            0x000000018ec897a4 _sigtramp + 56
[ 2] 1017a088c - initialize_gltf_module(ModuleInitializationLevel) (in godot.macos.editor.dev.arm64) + 104
[ 3] 1017a088c - initialize_gltf_module(ModuleInitializationLevel) (in godot.macos.editor.dev.arm64) + 104
[ 4] 1015995b8 - initialize_modules(ModuleInitializationLevel) (in godot.macos.editor.dev.arm64) + 144
[ 5] 1009e4d50 - Main::setup(char const*, int, char**, bool) (in godot.macos.editor.dev.arm64) + 18836
[ 6] 10092f788 - OS_MacOS_NSApp::start_main() (in godot.macos.editor.dev.arm64) + 60
[ 7] 10093b194 - -[GodotApplicationDelegate applicationDidFinishLaunching:] (in godot.macos.editor.dev.arm64) + 52
[ 8] 18ed3164c - 8   CoreFoundation                      0x000000018ed3164c __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 148
[ 9] 18edbdabc - 9   CoreFoundation                      0x000000018edbdabc ___CFXRegistrationPost_block_invoke + 92
[10] 18edbda00 - 10  CoreFoundation                      0x000000018edbda00 _CFXRegistrationPost + 440
[11] 18ed00170 - 11  CoreFoundation                      0x000000018ed00170 _CFXNotificationPost + 740
[12] 1905344a4 - 12  Foundation                          0x00000001905344a4 -[NSNotificationCenter postNotificationName:object:userInfo:] + 88
[13] 19316f5e8 - 13  AppKit                              0x000000019316f5e8 -[NSApplication _postDidFinishNotification] + 308
[14] 19316f380 - 14  AppKit                              0x000000019316f380 -[NSApplication _sendFinishLaunchingNotification] + 172
[15] 19316d9c4 - 15  AppKit                              0x000000019316d9c4 -[NSApplication(NSAppleEventHandling) _handleAEOpenEvent:] + 488
[16] 19316d5d8 - 16  AppKit                              0x000000019316d5d8 -[NSApplication(NSAppleEventHandling) _handleCoreEvent:withReplyEvent:] + 488
[17] 19055cfbc - 17  Foundation                          0x000000019055cfbc -[NSAppleEventManager dispatchRawAppleEvent:withRawReply:handlerRefCon:] + 316
[18] 19055cdb4 - 18  Foundation                          0x000000019055cdb4 _NSAppleEventManagerGenericHandler + 80
[19] 196f65f34 - 19  AE                                  0x0000000196f65f34 _AppleEventsCheckInAppWithBlock + 13900
[20] 196f6588c - 20  AE                                  0x0000000196f6588c _AppleEventsCheckInAppWithBlock + 12196
[21] 196f5efb8 - 21  AE                                  0x0000000196f5efb8 aeProcessAppleEvent + 484
[22] 19ba69344 - 22  HIToolbox                           0x000000019ba69344 AEProcessAppleEvent + 68
[23] 193167348 - 23  AppKit                              0x0000000193167348 _DPSNextEvent + 1284
[24] 193cfc69c - 24  AppKit                              0x0000000193cfc69c -[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 688
[25] 193cfc3a8 - 25  AppKit                              0x0000000193cfc3a8 -[NSApplication(NSEventRouting) nextEventMatchingMask:untilDate:inMode:dequeue:] + 72
[26] 19315a13c - 26  AppKit                              0x000000019315a13c -[NSApplication run] + 368
[27] 10092f740 - OS_MacOS_NSApp::run() (in godot.macos.editor.dev.arm64) + 32
[28] 1009dcad8 - main (in godot.macos.editor.dev.arm64) + 1560
[29] 18e8c3da4 - 29  dyld                                0x000000018e8c3da4 start + 6992
-- END OF C++ BACKTRACE --
================================================================
```

macOS crash without this PR for comparison:
```
================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.7.dev.custom_build (ee713ccb7cbaddf59f8258ed6663461779d00856)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
Load address: 1044dc000

[1] 18ec897a4 - 1   libsystem_platform.dylib            0x000000018ec897a4 _sigtramp + 56
[2] 105a4a1e0 - initialize_gltf_module(ModuleInitializationLevel) (in godot.macos.editor.dev.arm64) + 88
[3] 105a4a1e0 - initialize_gltf_module(ModuleInitializationLevel) (in godot.macos.editor.dev.arm64) + 88
[4] 105842f1c - initialize_modules(ModuleInitializationLevel) (in godot.macos.editor.dev.arm64) + 144
[5] 104c8e050 - Main::setup(char const*, int, char**, bool) (in godot.macos.editor.dev.arm64) + 18836
[6] 104bd9258 - OS_MacOS_NSApp::start_main() (in godot.macos.editor.dev.arm64) + 60
[7] 104be3e20 - -[GodotApplicationDelegate applicationDidFinishLaunching:] (in godot.macos.editor.dev.arm64) + 52
[8] 18ed3164c - 8   CoreFoundation                      0x000000018ed3164c __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 148
[9] 18edbdabc - 9   CoreFoundation                      0x000000018edbdabc ___CFXRegistrationPost_block_invoke + 92
[10] 18edbda00 - 10  CoreFoundation                      0x000000018edbda00 _CFXRegistrationPost + 440
[11] 18ed00170 - 11  CoreFoundation                      0x000000018ed00170 _CFXNotificationPost + 740
[12] 1905344a4 - 12  Foundation                          0x00000001905344a4 -[NSNotificationCenter postNotificationName:object:userInfo:] + 88
[13] 19316f5e8 - 13  AppKit                              0x000000019316f5e8 -[NSApplication _postDidFinishNotification] + 308
[14] 19316f380 - 14  AppKit                              0x000000019316f380 -[NSApplication _sendFinishLaunchingNotification] + 172
[15] 19316d9c4 - 15  AppKit                              0x000000019316d9c4 -[NSApplication(NSAppleEventHandling) _handleAEOpenEvent:] + 488
[16] 19316d5d8 - 16  AppKit                              0x000000019316d5d8 -[NSApplication(NSAppleEventHandling) _handleCoreEvent:withReplyEvent:] + 488
[17] 19055cfbc - 17  Foundation                          0x000000019055cfbc -[NSAppleEventManager dispatchRawAppleEvent:withRawReply:handlerRefCon:] + 316
[18] 19055cdb4 - 18  Foundation                          0x000000019055cdb4 _NSAppleEventManagerGenericHandler + 80
[19] 196f65f34 - 19  AE                                  0x0000000196f65f34 _AppleEventsCheckInAppWithBlock + 13900
[20] 196f6588c - 20  AE                                  0x0000000196f6588c _AppleEventsCheckInAppWithBlock + 12196
[21] 196f5efb8 - 21  AE                                  0x0000000196f5efb8 aeProcessAppleEvent + 484
[22] 19ba69344 - 22  HIToolbox                           0x000000019ba69344 AEProcessAppleEvent + 68
[23] 193167348 - 23  AppKit                              0x0000000193167348 _DPSNextEvent + 1284
[24] 193cfc69c - 24  AppKit                              0x0000000193cfc69c -[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 688
[25] 193cfc3a8 - 25  AppKit                              0x0000000193cfc3a8 -[NSApplication(NSEventRouting) nextEventMatchingMask:untilDate:inMode:dequeue:] + 72
[26] 19315a13c - 26  AppKit                              0x000000019315a13c -[NSApplication run] + 368
[27] 104bd9210 - OS_MacOS_NSApp::run() (in godot.macos.editor.dev.arm64) + 32
[28] 104c85dd8 - main (in godot.macos.editor.dev.arm64) + 1560
[29] 18e8c3da4 - 29  dyld                                0x000000018e8c3da4 start + 6992
-- END OF C++ BACKTRACE --
================================================================
```

Windows print_stack_trace non-verbose:
```
[ 1] main       initialize_gltf_module (register_types.cpp:113)
[ 2] main       initialize_modules (register_module_types.gen.cpp:146)
[ 3] main       Main::setup (main.cpp:2256)
[ 4] main       widechar_main (godot_windows.cpp:83)
[ 5] main       _main (godot_windows.cpp:125)
[ 6] main       main (godot_windows.cpp:139)
[ 7] main       WinMain (godot_windows.cpp:152)
```

Windows print_stack_trace in verbose mode:
```
[ 1] 7ff702a5b7c0 (main+39b7c0) initialize_gltf_module (C:\Users\aaronfranke\workspace\godot\modules\gltf\register_types.cpp:113)
[ 2] 7ff702911e4c (main+251e4c) initialize_modules (C:\Users\aaronfranke\workspace\godot\modules\register_module_types.gen.cpp:146)
[ 3] 7ff7028bd604 (main+1fd604) Main::setup (C:\Users\aaronfranke\workspace\godot\main\main.cpp:2256)
[ 4] 7ff702821868 (main+161868) widechar_main (C:\Users\aaronfranke\workspace\godot\platform\windows\godot_windows.cpp:83)
[ 5] 7ff702821a5c (main+161a5c) _main (C:\Users\aaronfranke\workspace\godot\platform\windows\godot_windows.cpp:125)
[ 6] 7ff702821a94 (main+161a94) main (C:\Users\aaronfranke\workspace\godot\platform\windows\godot_windows.cpp:139)
[ 7] 7ff702821628 (main+161628) WinMain (C:\Users\aaronfranke\workspace\godot\platform\windows\godot_windows.cpp:152)
```

Windows crash with this PR:
```
================================================================
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.7.dev.custom_build (cc0a112da8476b7b0651168a9d14cfb3d17f7f08)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
Load address: 7ff7026c0000

[0] 7ff7060f837c (main+3a3837c) - Node::set_name (C:\Users\aaronfranke\workspace\godot\scene\main\node.cpp:1442)
[1] 7ff7060f837c (main+3a3837c) - Node::set_name (C:\Users\aaronfranke\workspace\godot\scene\main\node.cpp:1442)
[2] 7ff702a5b7e8 (main+39b7e8) - initialize_gltf_module (C:\Users\aaronfranke\workspace\godot\modules\gltf\register_types.cpp:114)
[3] 7ff702911e4c (main+251e4c) - initialize_modules (C:\Users\aaronfranke\workspace\godot\modules\register_module_types.gen.cpp:146)
[4] 7ff7028bd604 (main+1fd604) - Main::setup (C:\Users\aaronfranke\workspace\godot\main\main.cpp:2256)
[5] 7ff702821868 (main+161868) - widechar_main (C:\Users\aaronfranke\workspace\godot\platform\windows\godot_windows.cpp:83)
[6] 7ff702821a5c (main+161a5c) - _main (C:\Users\aaronfranke\workspace\godot\platform\windows\godot_windows.cpp:125)
[7] 7ff702821a94 (main+161a94) - main (C:\Users\aaronfranke\workspace\godot\platform\windows\godot_windows.cpp:139)
[8] 7ff702821628 (main+161628) - WinMain (C:\Users\aaronfranke\workspace\godot\platform\windows\godot_windows.cpp:152)
[9] 7ff70a437af4 (main+7d77af4) - __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[10] 7ff70a437d6c (main+7d77d6c) - WinMainCRTStartup (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_winmain.cpp:16)
[11] 7ffce2c18740 (kernel32.dllôoè!ïòþ#+8740) - <couldn't map PC to fn name>
-- END OF C++ BACKTRACE --
================================================================
```

Windows crash without this PR for comparison:
```
================================================================
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.7.dev.custom_build (ee713ccb7cbaddf59f8258ed6663461779d00856)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
Load address: 7ff62e7d0000

[0] 7ff63220647c (main+3a3647c) - Node::set_name (C:\Users\aaronfranke\workspace\godot\scene\main\node.cpp:1442)
[1] 7ff62eb698fc (main+3998fc) - initialize_gltf_module (C:\Users\aaronfranke\workspace\godot\modules\gltf\register_types.cpp:113)
[2] 7ff62ea1ff6c (main+24ff6c) - initialize_modules (C:\Users\aaronfranke\workspace\godot\modules\register_module_types.gen.cpp:146)
[3] 7ff62e9cb724 (main+1fb724) - Main::setup (C:\Users\aaronfranke\workspace\godot\main\main.cpp:2256)
[4] 7ff62e92f9e0 (main+15f9e0) - widechar_main (C:\Users\aaronfranke\workspace\godot\platform\windows\godot_windows.cpp:83)
[5] 7ff62e92fbd4 (main+15fbd4) - _main (C:\Users\aaronfranke\workspace\godot\platform\windows\godot_windows.cpp:125)
[6] 7ff62e92fc0c (main+15fc0c) - main (C:\Users\aaronfranke\workspace\godot\platform\windows\godot_windows.cpp:139)
[7] 7ff62e92f7a0 (main+15f7a0) - WinMain (C:\Users\aaronfranke\workspace\godot\platform\windows\godot_windows.cpp:152)
[8] 7ff636545de4 (main+7d75de4) - __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[9] 7ff63654605c (main+7d7605c) - WinMainCRTStartup (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_winmain.cpp:16)
[10] 7ffce2c18740 (kernel32.dll+8740) - <couldn't map PC to fn name>
-- END OF C++ BACKTRACE --
================================================================
```

Note: The crashes were done using `Node *node = nullptr; node->set_name("crash");` to simulate a real crash.

The generic Unix code on macOS, which was used when I opened this PR, but is not actually used anymore because of the macOS-specific code, but if that code was removed, this would be the output:

```
2   godot.macos.editor.dev.arm64        0x0000000101f03288 _Z22initialize_gltf_module25ModuleInitializationLevel + 48
3   godot.macos.editor.dev.arm64        0x0000000101cfbfec _Z18initialize_modules25ModuleInitializationLevel + 144
4   godot.macos.editor.dev.arm64        0x0000000101147558 _ZN4Main5setupEPKciPPcb + 18836
5   godot.macos.editor.dev.arm64        0x0000000101091258 _ZN14OS_MacOS_NSApp10start_mainEv + 60
6   godot.macos.editor.dev.arm64        0x000000010109be20 -[GodotApplicationDelegate applicationDidFinishLaunching:] + 52
7   CoreFoundation                      0x000000018ed3164c __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 148
8   CoreFoundation                      0x000000018edbdabc ___CFXRegistrationPost_block_invoke + 92
9   CoreFoundation                      0x000000018edbda00 _CFXRegistrationPost + 440
10  CoreFoundation                      0x000000018ed00170 _CFXNotificationPost + 740
11  Foundation                          0x00000001905344a4 -[NSNotificationCenter postNotificationName:object:userInfo:] + 88
12  AppKit                              0x000000019316f5e8 -[NSApplication _postDidFinishNotification] + 308
13  AppKit                              0x000000019316f380 -[NSApplication _sendFinishLaunchingNotification] + 172
14  AppKit                              0x000000019316d9c4 -[NSApplication(NSAppleEventHandling) _handleAEOpenEvent:] + 488
15  AppKit                              0x000000019316d5d8 -[NSApplication(NSAppleEventHandling) _handleCoreEvent:withReplyEvent:] + 488
16  Foundation                          0x000000019055cfbc -[NSAppleEventManager dispatchRawAppleEvent:withRawReply:handlerRefCon:] + 316
17  Foundation                          0x000000019055cdb4 _NSAppleEventManagerGenericHandler + 80
18  AE                                  0x0000000196f65f34 _AppleEventsCheckInAppWithBlock + 13900
19  AE                                  0x0000000196f6588c _AppleEventsCheckInAppWithBlock + 12196
20  AE                                  0x0000000196f5efb8 aeProcessAppleEvent + 484
21  HIToolbox                           0x000000019ba69344 AEProcessAppleEvent + 68
22  AppKit                              0x0000000193167348 _DPSNextEvent + 1284
23  AppKit                              0x0000000193cfc69c -[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 688
24  AppKit                              0x0000000193cfc3a8 -[NSApplication(NSEventRouting) nextEventMatchingMask:untilDate:inMode:dequeue:] + 72
25  AppKit                              0x000000019315a13c -[NSApplication run] + 368
26  godot.macos.editor.dev.arm64        0x0000000101091210 _ZN14OS_MacOS_NSApp3runEv + 32
```

The generic Unix code on Linux, which was used when I opened this PR, but is not actually used anymore because of the Linux-specific code, but if that code was removed, this would be the output:

```
./bin/godot.linuxbsd.editor.dev.arm64(_Z22initialize_gltf_module25ModuleInitializationLevel+0x30) [0xb073801f4d2c]
./bin/godot.linuxbsd.editor.dev.arm64(_Z18initialize_modules25ModuleInitializationLevel+0x8c) [0xb0737ffe48c8]
./bin/godot.linuxbsd.editor.dev.arm64(_ZN4Main6setup2Eb+0x1244) [0xb0737f6595d4]
./bin/godot.linuxbsd.editor.dev.arm64(_ZN4Main5setupEPKciPPcb+0xd43c) [0xb0737f657f08]
./bin/godot.linuxbsd.editor.dev.arm64(main+0x114) [0xb0737f5cf2e0]
/lib/aarch64-linux-gnu/libc.so.6(+0x284c4) [0xe21f2ef584c4]
```

**Comment by AThousandShips** [comment]
I'll try to take a look to see if this erases recent improvements on Linux, but with the code shuffled around it's a bit hard to tell

File: platform/linuxbsd/stack_trace_linuxbsd.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+#include <link.h>
+#endif
+
+namespace StackTraceLinuxBSD {
+
+inline String select_addr2line_executable(OS *p_os) {
+	List<String> args;
+	args.push_back("--version");
+	if (p_os->has_environment("HOME")) {
+		const String cargo_addr2line = p_os->get_environment("HOME").path_join("/.cargo/bin/addr2line");
```
Comment (AThousandShips) [inline platform/linuxbsd/stack_trace_linuxbsd.h]: This appears to skip part of the original fetch for cargo, could you explain that change?

File: platform/linuxbsd/stack_trace_linuxbsd.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+inline Vector<String> symbolize_with_addr2line(OS *p_os, const String &p_addr2line_exe, const Vector<uint64_t> &p_addresses, uintptr_t p_relocation, const String &p_exec_path) {
+	List<String> args;
+	for (const uint64_t address_u64 : p_addresses) {
+		char address[32];
+		snprintf(address, sizeof(address), "%p", (void *)((uintptr_t)address_u64 - p_relocation));
+		args.push_back(address);
+	}
+	args.push_back("-e");
+	args.push_back(p_exec_path);
```
Comment (AThousandShips) [inline platform/linuxbsd/stack_trace_linuxbsd.h:78]: You are dropping several arguments here that are necessary:

```suggestion
	args.push_back(p_exec_path);
	args.push_back("-f");
	args.push_back("-p");
	args.push_back("-C");
```

**Review by AThousandShips — COMMENTED** [review]
Hard to tell the exact changes, but I'd suggest reimplementing it from scratch using the existing code to make sure nothing is missed, with the code reshuffled it's pretty hard to tell on my end

File: platform/linuxbsd/stack_trace_linuxbsd.h (hunk 1 above)
Comment (aaronfranke) [reply platform/linuxbsd/stack_trace_linuxbsd.h]: It doesn't skip, except when the `HOME` environment variable is undefined.

But I agree with you that it's hard to tell with this code shuffled around, so I opened another PR which is smaller and easier to review: PR https://github.com/godotengine/godot/pull/118642. Let's put this PR (#64205) on hold until PR #118642 is reviewed and merged.

**Comment by bruvzg** [comment]
It's definitely useful to have a way to print traces.

I would probably split it into platform specific function returning same data structure, and a common print method formatting this print (related draft - https://github.com/godotengine/godot/pull/91655).

Something like:
OS::get_stack_trace() -> TraceData
OS::print_stiack_trace(data, skip)

TraceData:
- Array of loaded modules: `name`, `load address`
- Array of stack trace frames: `module idx`, `address`, `info string` (ideally can be split into `function`/`file`/`line no`, but formatting of `add2line`/`atos` a pretty inconsistent so not sure if it's worth doing).

Some notes about current state:
- Linux changes seem to be missing recently added module info.
- Window code seems to only touch SEH handler, which is only used by MSVC builds and therefore not used by official builds. Relevant MinGW handler is in `crash_handler_windows_signal.cpp`. `print_stack_trace` also only use SEH code, which won't work for MinGW.

**Comment by almic** [comment]
I'm actually looking to track a bug related to node positions transforming in unrealistic ways. I'm mostly suspicious of interference from the physics server, but the best I can do is track the notification call `NOTIFICATION_LOCAL_TRANSFORM_CHANGED` at the moment it happens... but it's useless because there is no script backtrace. I set the transform in a single place in my code, and that isn't tripping because I know the modifications are always a fraction of a degree.

Basically, I would LOVE to have this available in GDscript purely to help confirm my suspicions and start following the thread back into the engine code. If I had this now, I would have a quick starting place to track backwards and see if it really is interference from physics.

(I don't need debugging help, and I already know for a fact I don't cause the unstable behavior, I'm just saying this is definitely a desired feature from the scripting side)
